tests/templates/test_subroutines/test_amplitude_amplification.py

```diff
@@ -0,0 +1,336 @@
+        ):
+            qml.AmplitudeAmplification(U, O, iters=3, fixed_point=True)
+
+    @pytest.mark.parametrize(
+        "wires, fixed_point, work_wire",
+        (
+            ([0, 1, 2], True, 2),
+            (["a", "b"], True, "a"),
+        ),
+    )
+    def test_error_wrong_work_wire(self, wires, fixed_point, work_wire):
+        """Test that an error is raised if work_wire is part of the O wires."""
+
+        U = generator(wires=wires)
+        O = oracle([0], wires=wires)
+
+        with pytest.raises(ValueError, match="work_wire must be different from the wires of O."):
+            qml.AmplitudeAmplification(U, O, iters=3, fixed_point=fixed_point, work_wire=work_wire)
+
+
+@pytest.mark.parametrize(
+    "n_wires, items, iters",
+    (
+        (3, [0, 2], 1),
+        (3, [1, 2], 2),
+        (5, [4, 5, 7, 12], 3),
+        (5, [0, 1, 2, 3, 4], 4),
+    ),
+)
+def test_compare_grover(n_wires, items, iters):
+    """Test that Grover's algorithm gives the same result with GroverOperator and AmplitudeAmplification."""
+    U = generator(wires=range(n_wires))
+    O = oracle(items, wires=range(n_wires))
+
+    dev = qml.device("default.qubit", wires=n_wires)
+
+    @qml.qnode(dev)
+    def circuit_amplitude_amplification():
+        generator(wires=range(n_wires))
+        qml.AmplitudeAmplification(U, O, iters)
+        return qml.probs(wires=range(n_wires))
+
+    @qml.qnode(dev)
+    def circuit_grover():
+        generator(wires=range(n_wires))
+
+        for _ in range(iters):
+            oracle(items, wires=range(n_wires))
+            qml.GroverOperator(wires=range(n_wires))
+
+        return qml.probs(wires=range(n_wires))
+
+    assert np.allclose(circuit_amplitude_amplification(), circuit_grover(), atol=1e-5)
+
+
+def test_default_lightning_devices():
+    """Test that AmplitudeAmplification executes with the default.qubit and lightning.qubit simulators."""
+
+    def circuit():
+        """Test circuit"""
+        qml.Hadamard(wires=0)
+        qml.Hadamard(wires=1)
+        qml.Hadamard(wires=2)
+
+        qml.AmplitudeAmplification(
+            generator(range(3)), oracle([0], range(3)), fixed_point=True, iters=3, work_wire=3
+        )
+        return qml.probs(wires=range(3))
+
+    dev1 = qml.device("default.qubit")
+    qnode1 = qml.QNode(circuit, dev1, interface=None)
+
+    res1 = qnode1()
+
+    """Test that the QNode executes with the Lightning Qubit simulator."""
+    dev2 = qml.device("lightning.qubit", wires=4)
+    qnode2 = qml.QNode(circuit, dev2)
+
+    res2 = qnode2()
+
+    assert np.allclose(res1, res2, atol=1e-5)
+
+
+class TestDifferentiability:
+    """Test that AmplitudeAmplification is differentiable"""
+
+    @staticmethod
+    def circuit(params):
+        qml.RY(params[0], wires=0)
+        qml.AmplitudeAmplification(
+            qml.RY(params[0], wires=0),
+            qml.RZ(params[1], wires=0),
+            iters=3,
+            fixed_point=True,
+            work_wire=3,
```

Comment:
This is outside of scope of this PR, but it would be really cool if we could optimize the iterations as well 😎

Reply:
depends on several factors but it would be very interesting 😄